### PR TITLE
feat(date-picker): add footer slot

### PR DESCRIPTION
**Pull request by timeTravelCYN**

### CODE REVIEW

File: src/date-picker/demos/enUS/footerslot.demo.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  <n-date-picker v-model:value="daterange" type="daterange">
+    <template #footer> extra footer </template>
+  </n-date-picker>
+  <n-date-picker v-model:value="datetimerange" type="datetimerange">
+    <template #footer> extra footer </template>
+  </n-date-picker>
+</n-space>
+```
+
+```js
```
Comment (07akioni) [inline src/date-picker/demos/enUS/footerslot.demo.md:20]: remove all v-model, use uncontrolled mode. it's clearer.

File: src/date-picker/demos/zhCN/footerslot.demo.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# 额外的页脚
```
Comment (07akioni) [inline src/date-picker/demos/zhCN/footerslot.demo.md:1]: 额外内容

File: src/date-picker/demos/enUS/index.demo-entry.md [hunk 3]
```diff
@@ -81,3 +82,9 @@ ranges
 | separator | `string` | `'to'` |  |
 | start-placeholder | `string` | `'Start Date and Time'` |  |
 | on-update:value | `(value: [number, number] \| null) => void` | `undefined` |  |
+
+## Slots
+
+| 名称   | 参数 | 说明             |
+| ------ | ---- | ---------------- |
+| footer | `()` | add Extra Footer |
```
Comment (07akioni) [inline src/date-picker/demos/enUS/index.demo-entry.md:90]:
```suggestion
| footer | `()` | Extra footer. |
```

File: src/date-picker/demos/zhCN/footerslot.demo.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    <template #footer> extra footer </template>
+  </n-date-picker>
+  <n-date-picker v-model:value="datetimerange" type="datetimerange">
+    <template #footer> extra footer </template>
+  </n-date-picker>
+</n-space>
+```
+
+```js
+export default {
```
Comment (07akioni) [inline src/date-picker/demos/zhCN/footerslot.demo.md:21]: remove v-model and related data

File: src/date-picker/src/DatePicker.tsx [hunk 5]
```diff
@@ -466,6 +466,9 @@ export default defineComponent({
       ...uniVaidation,
       ...dualValidation
     })
+    provide('datePickerSlot', {
```
Comment (07akioni) [inline src/date-picker/src/DatePicker.tsx:469]: Provide it using `datePickerInjectionKey`. There's already a provide in the file. Use that is enough.

File: src/date-picker/src/panel/date.tsx [hunk 6]
```diff
@@ -102,30 +103,35 @@ export default defineComponent({
...
-                  theme={mergedTheme.peers.Button}
-                  themeOverrides={mergedTheme.peerOverrides.Button}
-                  size="tiny"
-                  onClick={this.handleNowClick}
-                >
-                  {{ default: () => this.locale.now }}
-                </NButton>
-              ) : null}
-              {/** we don't need a confirm button for date picking */}
+          <div class={`${mergedClsPrefix}-date-panel-footer`}>
```
Comment (07akioni) [inline src/date-picker/src/panel/date.tsx:106]: Footer and actions are independant rows. Do not wrap them in a div. Use css grid. It's already defined in cssr files.

File: src/date-picker/src/panel/dateSlots.tsx [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { h, defineComponent, inject, Slots, renderSlot } from 'vue'
+import { useConfig } from '../../../_mixins'
+export default defineComponent({
```
Comment (07akioni) [inline src/date-picker/src/panel/dateSlots.tsx:3]: No need for the file. Implement it in panels.

File: src/date-picker/src/panel/daterange.tsx [hunk 8]
```diff
@@ -190,54 +191,57 @@ export default defineComponent({
...
-                  themeOverrides={mergedTheme.peerOverrides.Button}
-                  size="tiny"
-                  type="primary"
-                  disabled={this.isRangeInvalid}
-                  onClick={this.handleConfirmClick}
-                >
-                  {{ default: () => this.locale.confirm }}
-                </NButton>
-              ) : null}
+          <div class={`${mergedClsPrefix}-date-panel-footer`}>
```
Comment (07akioni) [inline src/date-picker/src/panel/daterange.tsx:194]: The same, use css grid to do layout.

File: src/date-picker/src/panel/datetime.tsx [hunk 9]
```diff
@@ -133,41 +134,46 @@ export default defineComponent({
...
-                  themeOverrides={mergedTheme.peerOverrides.Button}
-                  size="tiny"
-                  type="primary"
-                  disabled={this.isDateInvalid}
-                  onClick={this.handleConfirmClick}
-                >
-                  {{ default: () => this.locale.confirm }}
-                </NButton>
-              ) : null}
+          <div class={`${mergedClsPrefix}-date-panel-footer`}>
```
Comment (07akioni) [inline src/date-picker/src/panel/datetime.tsx:137]: the same

File: src/date-picker/src/panel/datetimerange.tsx [hunk 10]
```diff
@@ -248,54 +249,57 @@ export default defineComponent({
...
-                  size="tiny"
-                  type="primary"
-                  disabled={this.isRangeInvalid}
-                  onClick={this.handleConfirmClick}
-                >
-                  {{ default: () => this.locale.confirm }}
-                </NButton>
-              ) : null}
+          <div class={`${mergedClsPrefix}-date-panel-footer`}>
+            <DatePanelFooter />
```
Comment (07akioni) [inline src/date-picker/src/panel/datetimerange.tsx:253]: same

**Review by 07akioni — COMMENTED** [review]
Note that action is independent of footer. So actions.length should determine whether to show the footer.

File: src/date-picker/src/panel/datetimerange.tsx (hunk 10 above)
Comment (timeTravelCYN) [reply src/date-picker/src/panel/datetimerange.tsx:253]: ok,i'll refactor it later

File: src/date-picker/src/styles/index.cssr.ts [hunk 11]
```diff
@@ -329,6 +336,11 @@ export default c([
       width: 1px;
       background-color: var(--calendar-divider-color);
     `),
+    cB('date-panel-footer', {
+      lineHeight: '38px',
```
Comment (07akioni) [inline src/date-picker/src/styles/index.cssr.ts:340]:
```suggestion
```

File: src/date-picker/styles/_common.ts [hunk 12]
```diff
@@ -2,6 +2,7 @@ export default {
   itemSize: '24px',
   itemCellWidth: '38px',
   itemCellHeight: '32px',
+  panelExtraFooterPadding: '0px 12px',
```
Comment (07akioni) [inline src/date-picker/styles/_common.ts:5]:
```suggestion
  panelExtraFooterPadding: '8px 12px',
```

**Pull request merged**